internal/crypto/helpers.go

```diff
@@ -93,32 +93,39 @@ func ValidateRequestCerts(r *http.Request, certs []*tls.Certificate) error {
 		return errors.New("TLS not enabled")
 	}
 
-	if len(r.TLS.PeerCertificates) == 0 {
-		return errors.New("Client TLS certificate is required")
-	}
+	peerCertificate := r.TLS.PeerCertificates[0]
+
+	certID := HexSHA256(peerCertificate.Raw)
 
-	for _, peerCertificate := range r.TLS.PeerCertificates {
-		certID := HexSHA256(peerCertificate.Raw)
-		for _, cert := range certs {
-			// In case a cert can't be parsed or is invalid,
-			// it will be present in the cert list as 'nil'
-			if cert == nil {
-				// Invalid cert, continue to next one
-				continue
-			}
-
-			// Extensions[0] contains cache of certificate SHA256
-			if string(cert.Leaf.Extensions[0].Value) == certID {
-				if time.Now().After(cert.Leaf.NotAfter) {
-					return ErrCertExpired
-				}
-				// Happy flow, we matched a certificate
-				return nil
-			}
+	for _, cert := range certs {
+		// In case a cert can't be parsed or is invalid,
+		// it will be present in the cert list as 'nil'
+		if cert == nil {
+			// Invalid cert, continue to next one
+			continue
 		}
+
+		if cert.Leaf.IsCA && verifyCertAgainstCA(cert, peerCertificate) {
+			return nil
+		}
+	}
+
+	return errors.New("Certificate with SHA256 " + certID + " not allowed")
+}
+
+func verifyCertAgainstCA(caCert *tls.Certificate, peerCert *x509.Certificate) bool {
```

Comment:
move to internal/crypto?

Reply:
it's already in `internal/crypto/helpers.go`. do you mean something else?

Reply:
misread it